### PR TITLE
[JENKINS-72469] Avoid repeated tool downloads from misconfigured HTTP servers

### PR DESCRIPTION
**Pull request by MarkEWaite**
## [JENKINS-72469] Avoid repeated tool downloads from misconfigured HTTP servers

The Azul Systems content delivery network stopped providing the last-modified header in their URL responses.  They only provide the ETag header.

Add ETag support (based on the [ETag specification](https://httpwg.org/specs/rfc9110.html#field.etag)) to the Jenkins FilePath URL download method so that if ETag is provided, we use the ETag value.  If last-modified is provided and matches, we continue to honor it as well.

[JENKINS-72469](https://issues.jenkins.io/browse/JENKINS-72469) has more details.

The [community post](https://community.jenkins.io/t/job-stuck-on-unpacking-global-jdk-tool/11272) also includes more details.

### Testing done

* Automated test added to FilePathTest for code changes on the controller.  The automated test confirms that even without a last-modified value, the later downloads are skipped if a matching ETag is received.  The automated test also confirms that download is skipped if OK is received with a matching ETag.  No automated test was added to confirm download on the agent because that path is not tested by any of the other test automation of this class.
* Interactive test with the Azul Systems JDK installer on the controller.  I created a tool installer for the Azul JDK.  I verified that before this change it was downloaded each time the job was run.  I verified that after the change it was downloaded only once.  
* Interactive test with the Azul Systems JDK installer on an agent.  I created a tool installer for the Azul JDK.  I verified that before this change it was downloaded each time the job was run.  I verified that after the change it was downloaded only once.
* Interactive test on the controller with a file download from an NGINX web server confirmed that the tool is downloaded once and then later runs of the job did not download the file again.

### Proposed changelog entries

- Avoid repeated tool downloads from misconfigured HTTP servers.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java
Comment (mawinter69) [inline core/src/main/java/hudson/FilePath.java:987]: This could use Util.fixEmptyAndTrim to always make it null. This avoids the double check for null and empty at later points in time

File: core/src/main/java/hudson/FilePath.java
Comment (mawinter69) [inline core/src/main/java/hudson/FilePath.java:1034]: same here

File: core/src/main/java/hudson/FilePath.java
Comment (mawinter69) [inline core/src/main/java/hudson/FilePath.java:1036]: The tag can also be weak, in that case the etag sent is `W/"<value>" . So we need to remove a leading `W/` before removing the quotes

**Review by mawinter69 — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/FilePath.java
Comment (MarkEWaite) [reply core/src/main/java/hudson/FilePath.java]: Included in df05da5bb5bfde52278bb25f27f77f894a1fd2eb .  Thanks for review.  Could you review again?

File: core/src/main/java/hudson/FilePath.java
Comment (MarkEWaite) [reply core/src/main/java/hudson/FilePath.java]: Included in df05da5bb5bfde52278bb25f27f77f894a1fd2eb .  Thanks for review.  Could you review again?

File: core/src/main/java/hudson/FilePath.java
Comment (MarkEWaite) [reply core/src/main/java/hudson/FilePath.java]: I read the specification more carefully and realized that the original implementation was not handling the values as simply as it could.  Adjusted in df05da5bb5bfde52278bb25f27f77f894a1fd2eb .  Could you review again?

**Review by mawinter69 — APPROVED** [review]
Looks good now

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by MarkEWaite** [comment]
Thanks for the approval @timja.

I plan to merge this pull request shortly before I go to sleep tonight so that it will be included in the weekly build tomorrow. I'd like it in at least one weekly before we select the next LTS baseline.

**Pull request merged**
